[DAG] Improve known bits of Zext/Sext loads with range metadata

davemgreen opened the pull request:
This extends the known bits for extending loads which have range metadata, handling the range metadata on the original memory type, extending that to the correct BitWidth.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3612,32 +3612,42 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
+      else if (ISD::isSEXTLoad(Op.getNode()))
+        Known = Known0.sext(BitWidth);
+      else if (ISD::isEXTLoad(Op.getNode()))
+        Known = Known0.anyext(BitWidth);
```
Do you need to guard against FP type ext loads?

davemgreen replied:
Hmm. As far as I can tell fp loads will not have range data, and so if we have a anyext fp load then we should get a `Known(LD->getMemoryVT().getFixedSizeInBits()).anyext(Bitwidth)`, which should be fine.
@arsenm any reason you think this might need to be guarded? Thanks.

arsenm replied:
I'm not sure MMO metadata is properly handled when bit casts are folded into load/store

arsenm commented on llvm/test/CodeGen/AArch64/setcc_knownbits.ll:
```diff
@@ -21,9 +21,7 @@ define noundef i1 @logger(i32 noundef %logLevel, ptr %ea, ptr %pll) {
-; CHECK-NEXT:    ldrb w8, [x8]
-; CHECK-NEXT:    cmp w8, #0
-; CHECK-NEXT:    cset w0, ne
+; CHECK-NEXT:    ldrb w0, [x8]
```
Not enough test updates? I would expect a diff from the zext/sext/anyext cases

david-arm replied:
Perhaps it's possible to use the range metadata to write some explicit tests for sext/anyext?

davemgreen replied:
Yeah I need to add some better tests. Whilst this seems to come up a fair amount during C++ code (bools often have range metadata applied), it can be difficult to test is simple cases as the `sext(load !range)` is often already optimized before it creates the `load sext !range`. It might require C++ testing.

nikic commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3612,32 +3612,42 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
+      // MemoryVT().getSizeInBits()), which is truncated to the correct elt size
+      // if it is know. These are then extended to the original VT sizes below.
+      if (const MDNode *MD = LD->getRanges()) {
+        computeKnownBitsFromRangeMetadata(*MD, Known0);
```
I'm a bit concerned that this will result in miscompiles due to broken poison handling in SDAG, see 9deee6bffa9c331f46c68e5dd4cb4abf93dc0716. To be conservative, we could require !noundef.

arsenm replied:
But that patch already handles this? The range metadata wouldn't appear in the load node in the first place (unless some other code went out of its way to add it)

nikic replied:
Oh yeah, I confused myself here.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -3612,32 +3612,42 @@ KnownBits SelectionDAG::computeKnownBits(SDValue Op, const APInt &DemandedElts,
-      if (LD->getExtensionType() == ISD::NON_EXTLOAD) {
+    } else if (Op.getResNo() == 0) {
+      KnownBits Known0(!LD->getMemoryVT().isScalableVT()
+                           ? LD->getMemoryVT().getSizeInBits()
```
nit: Can you call `getFixedSizeInBits` here instead given we know it's fixed-width? I think it's a bit clearer and it does have a compile-time advantage too because it avoids going through the more expensive TypeSize->uint64_t cast operator, which is a function call.

davemgreen committed:
[DAG] Improve known bits of Zext/Sext loads with range metadata

This extends the known bits for extending loads which have range metadata,
handling the range metadata on the original memory type, extending that to the
correct BitWidths.

davemgreen commented:
I've added a couple of C++ tests to make sure this is tested. Thanks.

davemgreen committed:
Use getFixedSizeInBits

david-arm reviewed (commented):
Thanks for the changes @davemgreen! The patch LGTM. There was just one comment about FP extending loads. I left a comment about possibly adding an assert - do you think this it's worth doing?

arsenm reviewed (approved):
would be good to get the equivalent in globalisel
